python/python/lance/dataset.py

```diff
@@ -689,6 +689,83 @@ def cleanup_old_versions(
             td_to_micros(older_than), delete_unverified
         )
 
+    def create_scalar_index(
+        self,
+        column: Union[str, List[str]],
+        index_type: str,
+        name: Optional[str] = None,
+        replace: bool = True,
+    ) -> LanceDataset:
+        """Create scalar index on column.
+
+        **Experimental API**
+
+        Parameters
+        ----------
+        column : str
+            The column to be indexed.  Must be a boolean, integer, float,
+            or string column.
+        index_type : str
+            The type of the index.  Only ``"BTREE"`` is supported now.
+        name : str, optional
+            The index name. If not provided, it will be generated from the
+            column name.
+        replace : bool, default True
+            Replace the existing index if it exists.
+
+        Examples
+        --------
+
+        .. code-block:: python
+
+            import lance
+
+            dataset = lance.dataset("/tmp/images.lance")
+            dataset.create_index(
+                "category",
+                "BTREE"
+            )
+
+        Experimental Status:
+
+        Scalar indices are experimental and only utilized if there is no new or deleted
+        data that has been uploaded since the index was created.  This limitation should
+        be addressed soon.  In the meantime, ``compact_files`` can be used to remove
+        deletion files and ``optimize_indices`` can be used to catch up the index with
+        new data.
+        """
+        if isinstance(column, str):
+            column = [column]
+
+        if len(column) > 1:
+            raise NotImplementedError(
+                "Scalar indices currently only support a single column"
+            )
+
+        column = column[0]
+        if column not in self.schema.names:
+            raise KeyError(f"{column} not found in schema")
+
+        field = self.schema.field(column)
+        if (
+            not pa.types.is_integer(field.type)
+            and not pa.types.is_floating(field.type)
+            and not pa.types.is_boolean(field.type)
+            and not pa.types.is_string(field.type)
+        ):
+            raise TypeError(
+                f"Scalar index column {column} must be int, float, bool, or str"
+            )
+
+        index_type = index_type.upper()
+        if index_type != "BTREE":
+            raise NotImplementedError(
+                f"Only 'BTREE' is supported for index_type.  Received {index_type}"
+            )
+
+        self._ds.create_index([column], index_type, name, replace)
+        return LanceDataset(self.uri)
```

Comment:
This clears the session cache, which isn't ideal. It would be nice if `self._ds.create_index()` mutated `self` to update the dataset reference.

Reply:
In fact, this *is* already handled here:

https://github.com/lancedb/lance/blob/5a510afb182d772301622bab90e974fc2baeb7ab/python/src/dataset.rs#L707

Reply:
I changed this to return nothing (since it is already updating `self.ds` as you pointed out).  We should probably fix the other `create_index` function too (although we can defer that for a different PR)